src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -1362,74 +1378,74 @@ static bool CopyFile_CopyFileRange(int inFd, int outFd, const struct stat_ *sour
 }
 #endif // defined(PAL_COPY_FILE_RANGE_SYSCALL)
 
-int32_t SystemNative_CopyFile(intptr_t sourceFd, const char* srcPath, const char* destPath, int32_t overwrite)
+int32_t SystemNative_CopyFile(const char* srcPath, const char* destPath, int32_t overwrite)
 {
-    int inFd = ToFileDescriptor(sourceFd);
-    int outFd;
+    intptr_t inFd;
+    intptr_t outFd;
     int ret;
     int tmpErrno;
     struct stat_ sourceStat;
+    bool locked = false;
     bool copied = false;
 
-    while ((ret = fstat_(inFd, &sourceStat)) < 0 && errno == EINTR);
-    if (ret != 0)
+    // Not all file copying methods require an open file descriptor for the source file; however,
+    // to provide a behavior similar to Windows, we maintain POSIX advisory locks while the file is
+    // being copied.
+    inFd = SystemNative_Open(srcPath, PAL_O_RDONLY | PAL_O_CLOEXEC, 0);
+    if (inFd < 0)
     {
-        return -1;
+        goto out_no_open;
     }
 
-    struct stat_ destStat;
-    while ((ret = stat_(destPath, &destStat)) < 0 && errno == EINTR);
-    if (ret == 0)
+    if (SystemNative_FLock(inFd, LOCK_EX | LOCK_NB) < 0)
```

Comment:
Should not this be unlocked somewhere later?

Reply:
No. See comment down below: when the file is closed, the lock is dropped.

Reply:
Fair point. The man agrees with that although I remember having to do that manually on macOS when I was testing my original PR 🤷‍♂️. I was trying to figure out why some of the unit tests are still failing.

Reply:
I wouldn't be surprised, though, if macOS had a, uh, unique behavior here.  Wouldn't be the first time I encounter such things.

`flock()` operates on file descriptions rather than file descriptors, and since the description here is locked exclusively (as opposed to a shared lock), no other thread should be able to obtain that lock.  So even if another thread successfully opens the file, the `flock()` call should fail (and closing the file descriptor when this call fails will only drop the file description reference once, leaving the ref for the locked descriptor to close).  None of the tests are duplicating file descriptions, so even if another thread opens the source file again with `open()` this should be an issue.  However, if `flock()` on macOS behaves differently, treating a second `open()` in a file path as if it were a `dup()` on an already open/locked file, then things will be hairy to fix...

However,  if I'm reading this right, these failures seem to be related with not being able to upload results to Helix.  Let me try rebasing and pushing again to trigger a rebuild.  Let's see what happens from there.

Reply:
> However, if I'm reading this right, these failures seem to be related with not being able to upload results to Helix. Let me try rebasing and pushing again to trigger a rebuild. Let's see what happens from there.

Nah, the test failures were relevant:
https://dev.azure.com/dnceng/public/_build/results?buildId=519601&view=ms.vss-test-web.build-test-results-tab

Reply:
All the test failures are some form of `The process cannot access the file 'xxx' because it is being used by another process.`

Reply:
Just force-pushed a new revision that explicitly calls `flock(LOCK_UN)` before closing the file. Let's see what the bots will tell us.

Reply:
Runtime-live-build seems to have passed on macOS. :)

Reply:
The errors are coming only now :-/ It may take a while until all of them crop up with stack traces on Azure Pipelines test dashboard.

Reply:
Yeah. They're probably going to be the same as before. Will try to carve some time for this PR sometime this or the next week.

Reply:
I'll try to reproduce it tomorrow on my macOS machine and report back.